_applications/bad-apples.adoc

```diff
@@ -0,0 +1,310 @@
+unknown principal components for the transactions, we have taken these as
+inspiration to demonstrate the types of algorithms and rules you will create
+with your data.
+
+The techniques and algorithms demonstrated in this tutorial are **not**
+recommended for production work where real financial records are involved. They
+are exposed here to illustrate how the underlying infrastructure and
+frameworks can be used to build these type of systems.
+
+[[architecture]]
+== Architecture
+
+There are three primary services which comprise the Bad Apples suite: a
+database, a Spark cluster, and the filter
+https://docs.okd.io/latest/dev_guide/jobs.html[Job]. To aid this tutorial there are
+two additional services you will use: one to load the database, and one to
+observe the results of the filter processing.
+
+The filter job is an application that will extract credit card transaction
+data from the database, filter out transactions above a predefined amount, and
+then process the filtered transactions through the business rules to identify
+suspected fraudulent transactions. The business rules are crafted to select
+transactions based on one of the principal component values.
+
+In practice, the flow of data looks like this:
+
+pass:[<img src="/assets/howdoi/bad-apples.svg" alt="Bad Apples architecture" class="img-responsive">]
+
+Of the two non-primary services, the watcher is shown in this diagram to
+illustrate its place in the architecture. The watcher is a simple service
+that periodically reads from the database to display the number of detected
+fraudlent transactions. Using this service you can easily monitor how a change
+to the filter job affects the number of detected transactions.
+
+=== Record selection process
+
+The filter job service will examine all the records in the database and use
+a few simple algorithms to select the fraudulent transaction records. To better
+understand this process, let's begin by looking at the schema for the
+credit card transaction data.
+
+....
+CREATE TABLE transactions
+(
+Time VARCHAR,
+V1 VARCHAR,
+V2 VARCHAR,
+V3 VARCHAR,
+V4 VARCHAR,
+V5 VARCHAR,
+V6 VARCHAR,
+V7 VARCHAR,
+V8 VARCHAR,
+V9 VARCHAR,
+V10 VARCHAR,
+V11 VARCHAR,
+V12 VARCHAR,
+V13 VARCHAR,
+V14 VARCHAR,
+V15 VARCHAR,
+V16 VARCHAR,
+V17 VARCHAR,
+V18 VARCHAR,
+V19 VARCHAR,
+V20 VARCHAR,
+V21 VARCHAR,
+V22 VARCHAR,
+V23 VARCHAR,
+V24 VARCHAR,
+V25 VARCHAR,
+V26 VARCHAR,
+V27 VARCHAR,
+V28 VARCHAR,
+Amount VARCHAR,
+Class VARCHAR
+);
+....
+
+Each record in this set consists of a large amount of data that is mostly
+hidden in the
+https://en.wikipedia.org/wiki/Principal_component_analysis[principal components]
+V1 through V28. Due to confidentiality constraints, none of the original
+background information is available about those fields. The fields which have
+not been transformed with
+https://en.wikipedia.org/wiki/Principal_component_analysis[PCA] are `Time` and
+`Amount`. `Time` is the amount of time in seconds that has elapsed between
```

Comment:
I'm not clear what 'time' is. Is time for a specific transaction the time between that transaction and the earliest transaction in the data set? If so, maybe change to "that has elapsed between this transaction and ..." ?

Reply:
good callout, it is a weird value. so, it is the time in seconds that have elapsed between the current transaction and the first transaction in the data set. i'm going to rewrite this sentence.